builtin/gc.c

```diff
@@ -1615,15 +1615,46 @@ static int launchctl_remove_plists(const char *cmd)
 		launchctl_remove_plist(SCHEDULE_WEEKLY, cmd);
 }
 
+static int launchctl_list_contains_plist(const char *name, const char *cmd)
+{
+	int result;
+	struct child_process child = CHILD_PROCESS_INIT;
+	char *uid = launchctl_get_uid();
+
+	strvec_split(&child.args, cmd);
+	strvec_pushl(&child.args, "list", name, NULL);
+
+	child.no_stderr = 1;
+	child.no_stdout = 1;
+
+	if (start_command(&child))
+		die(_("failed to start launchctl"));
+
+	result = finish_command(&child);
+
+	free(uid);
+
+	/* Returns failure if 'name' doesn't exist. */
+	return !result;
+}
+
 static int launchctl_schedule_plist(const char *exec_path, enum schedule_priority schedule, const char *cmd)
 {
 	FILE *plist;
+	struct stat st;
 	int i;
 	const char *preamble, *repeat;
 	const char *frequency = get_frequency(schedule);
 	char *name = launchctl_service_name(frequency);
 	char *filename = launchctl_service_filename(name);
 
+	/* Does this file already exist? */
+	if (!stat(filename, &st) && st.st_size) {
+		/* Check if it is registered */
+		if (launchctl_list_contains_plist(name, cmd))
+			goto cleanup;
+	}
+
 	if (safe_create_leading_directories(filename))
 		die(_("failed to create directories for '%s'"), filename);
 	plist = xfopen(filename, "w");
```

Comment:
I would feel more comfortable if we called `hold_lock_file_for_update_timeout(&lock, filename, timeout_ms)` here.

And once the lock was acquired, not only write the file (i.e. do the very same stuff we do until that `fclose(plist)`, which would now be a `close_lock_file_gently(&lock)`), but _then_ verify that the plist is registered, and _then_ verify that the file contents are the same, and only if all those conditions hold true, `rollback_lock_file(&lock)` and `goto cleanup`, otherwise `commit_lock_file(&lock)` and do the `launchctl` dance.

What do you think?

Reply:
Since you indicated privately that you'd be fine with me implementing this, I went ahead and force-pushed.